src/Features/LanguageServer/Protocol/Extensions/Extensions.cs

```diff
@@ -21,12 +21,27 @@ namespace Microsoft.CodeAnalysis.LanguageServer
 {
     internal static class Extensions
     {
+        private const string SourceGeneratedFileScheme = "source-generated";
+        private const string SourceGeneratedGuidFormat = "D";
+
         public static Uri GetURI(this TextDocument document)
         {
             Contract.ThrowIfNull(document.FilePath);
-            return document is SourceGeneratedDocument
-                ? ProtocolConversions.GetUriFromPartialFilePath(document.FilePath)
-                : ProtocolConversions.GetUriFromFilePath(document.FilePath);
+
+            // If this is not a source generated document, we'll just use the URI stored in the FilePath
+            if (document is not SourceGeneratedDocument)
+                return ProtocolConversions.GetUriFromFilePath(document.FilePath);
+
+            // For source generated documents, we'll produce a URI specifically for LSP that has a scheme the client can register for; the "host" portion will
+            // just be the project ID of the document, then the path will be the GUID that is the project ID, with the rest of it being the generated file path
+            // just so any display of the end of the URI (like a file tab) works well.
+            return new Uri(
+                SourceGeneratedFileScheme + "://" +
+                document.Id.ProjectId.Id.ToString(SourceGeneratedGuidFormat) +
+                "/" +
+                document.Id.Id.ToString(SourceGeneratedGuidFormat) +
+                "/" +
+                document.FilePath);
```

Comment:
if only interpoalted strings were a thing :)

Reply:
you can also do nifty stuff like `{document.Id.Id:SourceGeneratedGuidFormat}` :)

Reply:
As best I can tell, there's no way to pass the constant as the format here -- your example does .ToString("SourceGeneratedGuidFormat") which isn't what I want here.